File: network/Filter.py
import math
from network.Matrix import Matrix
from network.WeightedNeuron import WeightedNeuron
from network import constants

class Filter:

    def __init__(self, activation = None, numNeurons = 9):
        self.activation = activation
        width = int(math.sqrt(numNeurons))
        self.neurons = Matrix([width, width])
        self.previousInput = None
        self.prevResult = None

        for i in range(0, width):
            for j in range(0, width):
                self.neurons[i][j] = WeightedNeuron(constants.multiply, activation, weightsDim=[1,1])

    def adjustNeuronWeights(self, deltas):
        assert len(deltas) == self.neurons.width() * self.neurons.height()

        for y in range(0, self.neurons.height()):
            for x in range(0, self.neurons.width()):
                index = x + (y * self.neurons.width())
                self.neurons[y][x].adjustWeights(deltas[index])

    def process(self, input):
        self.previousInput = input
        result = Filter.convolve(input, self.neurons)
        if self.activation != None:
            result = self.activation(result)

        self.prevResult = result
        return result

    @staticmethod
    def convolve(input: Matrix, feature: Matrix):
        """
        Convolves the input matrix with the given feature matrix.
        Returns the convolution as an output matrix (smaller in size from input)
        """
        outputWidth = input.width() - feature.width() + 1
        outputHeight = input.height() - feature.height() + 1

        # output = [[0 for x in range(len(input) - len(feature) + 1)] for y in range(len(input[0]) - len(feature[0]) + 1)]
        output = Matrix([outputWidth, outputHeight])
        # denom = len(feature) * len(feature[0])
        featureDim = feature.size()
        denom = featureDim[0] * featureDim[1]

        for i in range(outputHeight):
            for j in range(outputWidth):
                sum = 0

                for x in range(feature.height()):
                    for y in range(feature.width()):
                        sum += feature[x][y].multiply(input[x+i][y+j])[0][0]

                output[i][j] = sum / denom
        return output

File: java/src/Filter.java
import java.util.*;
import java.util.concurrent.Callable;

//import static org.junit.Assert.*;

class Filter {

    private Callable<Object> activation;
    private final Matrix<WeightedNeuron> neurons;
    private Object previousInput;
    private Object prevResult;

    public Filter() {
        this(null, 9);
    }

    public Filter(Callable<Object> activation, int numNeurons) {
        this.activation = activation;
        int width = (int)Math.sqrt(numNeurons);
        this.neurons = new Matrix(width, width);
        this.previousInput = null;
        this.prevResult = null;

        for (int i = 0; i < width; i++) {
            for (int j = 0; j < width; j++) {
                this.neurons.set(i, j, new WeightedNeuron(constants.multiply, activation, weightsDim =[1, 1]));
            }
        }
    }

    public void adjustNeuronWeights(List<Double> deltas) {

//        assertEqual(deltas.size(), this.neurons.width() * this.neurons.height());

        for (int y = 0; y < this.neurons.height(); y++) {
            for (int x = 0; x < this.neurons.width(); x++) {
                int index = x + (y * this.neurons.width());
                this.neurons.get(y, x).adjustWeights(deltas.get(index));
            }
        }
    }

    public Object process(Object input) {
        this.previousInput = input;
        Object result = Filter.convolve(input, this.neurons);
        if (this.activation != null) {
            result = this.activation.call(result);
        }

        this.prevResult = result;
        return result;
    }

    /**
     * Convolves the input matrix with the given feature matrix.
     * Returns the convolution as an output matrix (smaller in size from input)
     */
    private static Object convolve(Matrix<Integer> input, Matrix<WeightedNeuron> feature) {

        int outputWidth = input.width() - feature.width() + 1;
        int outputHeight = input.height() - feature.height() + 1;

        Matrix<Integer> output = new Matrix(outputWidth, outputHeight);
        int[] featureDim = feature.size();
        int denom = featureDim[0] * featureDim[1];

        for (int i = 0; i < outputHeight; i++) {
            for (int j = 0; j < outputWidth; j++) {
                int sum = 0;

                for (int x = 0; x < feature.height(); x++) {
                    for (int y = 0; y < feature.width() ; y++) {
                        sum += Matrix.multiply(input.get(x + i, y + j), feature.get(x, y))[0][0];
                    }
                }

                output.set(i, j , sum / denom);
            }
        }

        return output;
    }
}